contracts/core/OperationExecutor.sol

```diff
@@ -120,16 +126,74 @@ contract OperationExecutor is IERC3156FlashBorrower {
     bytes calldata data
   ) external override returns (bytes32) {
     address lender = registry.getRegisteredService(FLASH_MINT_MODULE);
+    FlashloanData memory flData = abi.decode(data, (FlashloanData));
+
+    checkIfLenderIsTrusted(lender);
+    checkIfFlashloanedAssetIsTheRequiredOne(asset, flData.asset);
+    checkIfFlashloanedAmountIsTheRequiredOne(asset, flData.amount);
+
+    processFlashloan(flData, initiator);
+
+    uint256 paybackAmount = amount.add(fee);
+    require(
+      IERC20(asset).balanceOf(address(this)) >= paybackAmount,
+      "Insufficient funds for payback"
+    );
 
-    require(msg.sender == lender, "Untrusted flashloan lender");
+    IERC20(asset).safeApprove(lender, paybackAmount);
+    return keccak256("ERC3156FlashBorrower.onFlashLoan");
+  }
 
+  function receiveFlashLoan(
+    IERC20[] memory tokens,
+    uint256[] memory amounts,
+    uint256[] memory feeAmounts,
+    bytes memory data
+  ) external override {
+    address asset = address(tokens[0]);
+    address lender = registry.getRegisteredService(BALANCER_VAULT);
     FlashloanData memory flData = abi.decode(data, (FlashloanData));
 
-    require(IERC20(asset).balanceOf(address(this)) >= flData.amount, "Flashloan inconsistency");
+    checkIfLenderIsTrusted(lender);
+    checkIfFlashloanedAssetIsTheRequiredOne(asset, flData.asset);
+    checkIfFlashloanedAmountIsTheRequiredOne(asset, flData.amount);
 
-    if (flData.isProxyFlashloan) {
-      IERC20(asset).safeTransfer(initiator, flData.amount);
+    // Can we just assume that the initiator is the OperationExecutor?
+    // Are there any security consequences because of that assumption?
```

Comment:
Not sure if we have an alternative with Balancer anyway.

Reply:
WEll that's the thing. I just assumed that it will be always the Executor,. IF you can't think of any security issues right there then it's fine I guess :D